refactor: migrate to starflow (#5225)
canonical/snapcraft: HACKING.md

Setup the environment by running:

```shell
make setup
```

### Tooling

We use a large number of tools for our project. Most of these are installed for

> you with `make setup`, but you'll need to install Python 3.12 separately.

Reviewer: Should we specify APT here, too?

If `make setup` throws without it, it feels safe to say that it's a requirement, and anyone who wants to develop without it must replicate the installation on their own.

Author: I think calling apt a "requirement" feels a bit weird to me. I suppose, strictly speaking, it is a requirement to use our automated "setup" step, but it isn't strictly a requirement to (for example) run any of the tests as once your environment is already set up, you'll never touch it.

It's also worth noting that this PR doesn't introduce anything new for apt to install. There were always "ghost" system package dependencies that you just had to trial & error your way through installing in order to get all of the tests to run.

Curious what others think, though.

Reviewer 2: You don't need apt, it's just that if you don't have apt you have to install the packages for your development environment yourself. So not a requirement, but a convenience. That's also why we use the warning function (not error) if we don't detect certain packages but you don't have apt.

Reviewer 2: You don't need apt, it's just that if you don't have apt you have to install the packages for your development environment yourself. So not a requirement, but a convenience. That's also why we use the warning function (not error) if we don't detect certain packages but you don't have apt.

### Testing

> For unit testing, you can use `make` recipes:

Reviewer: I think we should push for standard testing.

```suggestion
For unit testing, use the provided make recipes:
```

```shell
make test  # All unit tests
make test-fast  # Only the fast tests
make test-slow  # Only the slow tests
```

> For integration testing, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel. In order to use spread, first build Snapcraft using:

Reviewer: ```suggestion
For compatibility testing across supported platforms, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel.

To test with Spread, first build Snapcraft into a snap:
```

Reviewer: Ignore.

> For integration testing, Snapcraft uses [Spread](https://github.com/canonical/spread). Spread is a system for distributing tests and executing them in different backends, in parallel. In order to use spread, first build Snapcraft using:

Author: Spread tests are super slow to run, I think encouraging them every time might make the effort required to contribute artificially high. I'm a fan of your two-section approach, what do you think @mr-cal?

```shell
snapcraft pack
```

Then, move the resulting snap into the tests directory:

```shell
mv *.snap tests/
```

Next, install Spread with Go:

> ```

Reviewer: ```suggestion
```shell
```

go install github.com/snapcore/spread/cmd/spread@latest
```

> Note: Ensure that the `$HOME/go/bin` directory is on your path after installation.

Reviewer: ```suggestion
Ensure that the installation added the `$HOME/go/bin` directory to the `$PATH` environment variable.
```

> Then, you can run the integration tests using a local LXD backend with:

Reviewer: ```suggestion
Then, you can run the compatibility tests using a local LXD backend with:
```

Reviewer: Ignore.

> ```

Reviewer: ```suggestion
```shell
```

spread -v lxd:
```

> Or, you can run them in Google Cloud if you have a `SPREAD_GOOGLE_KEY`, with:

Reviewer: I think we should provide a link to a doc that describes the key – it's not evident what this is or how to get it.

Reviewer: Let's not do a link, then, but a quick phrase in that sentence describing it as Google Cloud credentials.
